it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java

```diff
@@ -0,0 +1,363 @@
+import static com.google.cloud.teleport.it.jdbc.JDBCResourceManagerUtils.generatePassword;
+
+import com.google.cloud.teleport.it.testcontainers.TestContainerResourceManager;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.repackaged.core.org.apache.commons.lang3.math.NumberUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.JdbcDatabaseContainer;
+
+/**
+ * Abstract class for implementation of {@link JDBCResourceManager} interface.
+ *
+ * <p>The class supports one database, and multiple tables per database object. A database is
+ * created when the container first spins up, if one is not given.
+ *
+ * <p>The database name is formed using testId. The database name will be "{testId}-{ISO8601 time,
+ * microsecond precision}", with additional formatting.
+ *
+ * <p>The class is thread-safe.
+ */
+public abstract class AbstractJDBCResourceManager<T extends JdbcDatabaseContainer<?>>
+    extends TestContainerResourceManager<JdbcDatabaseContainer<?>> implements JDBCResourceManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractJDBCResourceManager.class);
+
+  protected static final String DEFAULT_JDBC_USERNAME = "root";
+
+  protected final JDBCDriverFactory driver;
+  protected final String databaseName;
+  protected final String username;
+  protected final String password;
+
+  private final Map<String, String> tableIds;
+
+  @VisibleForTesting
+  AbstractJDBCResourceManager(
+      T container, AbstractJDBCResourceManager.Builder<T> builder, JDBCDriverFactory driver) {
+    super(
+        container
+            .withUsername(builder.username)
+            .withPassword(builder.password)
+            .withDatabaseName(builder.databaseName),
+        builder);
+
+    this.databaseName = container.getDatabaseName();
+    this.username = container.getUsername();
+    this.password = container.getPassword();
+    this.tableIds = new HashMap<>();
+    this.driver = driver;
+  }
+
+  protected AbstractJDBCResourceManager(
+      T container, AbstractJDBCResourceManager.Builder<T> builder) {
+    this(container, builder, new JDBCDriverFactory());
+  }
+
+  /**
+   * Return the default port that this JDBC implementation listens on.
+   *
+   * @return the JDBC port.
+   */
+  protected abstract int getJDBCPort();
+
+  @Override
+  public String getUsername() {
+    return username;
+  }
+
+  @Override
+  public String getPassword() {
+    return password;
+  }
+
+  @Override
+  public synchronized String getUri() {
+    return String.format(
+        "jdbc:%s://%s:%d/%s",
+        getDockerImageName().split(":")[0],
```

Comment:
Maybe we can create a method `getJdbcUrlPrefix` that can be overriden?

Reply:
Done.